Tinkers' Material and Traits,Support of "Ghost Items" and Custom Enchantments

kindlich opened the pull request:
- Builder for Tinkers' Construct Materials and Traits
[https://gist.github.com/kindlich/7dae66cf3fc7ab654ee5f48e4c1485cb](https://gist.github.com/kindlich/7dae66cf3fc7ab654ee5f48e4c1485cb)

- If the CoT item and fluid bracket handler would find nothing, it will now return an item representation that will try to update itself if queried, so that some item calls can be done even from within CoT scripts.

- Custom Enchantments, should be useable now.
<https://gist.github.com/kindlich/b1b070d15c26f626d65c176a4b4786c3>
They break  for example if the domain changes, then they will not be useable in that world any longer (But this shouldn't ever happen unless someone tweaks the scripts anyway.)

SkySom committed:
Merge branch 'develop/1.12' into release/1.12

SkySom committed:
Proper version for release!!!!

kindlich committed:
Added back TiCon and JEI and updated Forge

We need TiCon in there for the modifiers and TiCon wanted a newer Forge version

kindlich committed:
Added support for "Ghost" items and liquids which will update themselves asap

They will be fake stacks of either an empty item or a fluidstack and call the update method whenever modified to try to get the actual item.

kindlich committed:
Started with TiCon materials and Traits

Removed the current classes as they were unused.

kindlich committed:
Added damage and amount transfer to Ghost items

kindlich committed:
More Properties on Material and Trait builders, Refactured Functions class

kindlich committed:
More Work on ghost items

Now return themselves upon modification to check for calls on getInternal()

kindlich committed:
Added missing remove methods to MAterialBuilders and removed unused property

SkySom commented on src/main/java/com/teamacronymcoders/contenttweaker/modules/tinkers/brackets/BracketHandlerTiConTrait.java:
```diff
@@ -0,0 +1,34 @@
+    @Override
+    public IZenSymbol resolve(IEnvironmentGlobal environment, List<Token> tokens) {
+        if (tokens == null || tokens.size() < 3 || !tokens.get(0).getValue().equalsIgnoreCase("ticontrait"))
+            return null;
```
Always Brackets. Always

kindlich committed:
Fixed Ghost item updating

kindlich committed:
Added CanApplyTogetherEnchantment method to TiCon Traits

kindlich committed:
Added custom Enchantments

- Removed IEnchantment as CrT alredy has IEnchantmentDefinition
- Uses a builder to create the object.

Example:
<https://gist.github.com/kindlich/b1b070d15c26f626d65c176a4b4786c3>

Note:
Works fine during both, the CoT and the CrT loader.

kindlich committed:
All the braces {}

kindlich committed:
Added TraitRepresentation to the Trait functions

Should allow for more flexibility

kindlich committed:
Stuff I came accross while documenting

SkySom reviewed (approved):
Works for me.
